### PR TITLE
[AC-2567] Billing Performance Improvements

### PR DESCRIPTION
**Pull request by cturnbull-bitwarden**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/AC-2567

## 📔 Objective

The intention behind this PR is to improve some of the performance issues we're seeing with organizations that have a high number of invoices. The main gist of the issue lies with the fact that we're currently retrieving all invoices for a given subscriber and then displaying them on the page.  This becomes an issue for customers that have thousands, and sometimes tens of thousands of invoices.

There's also some cleanup in this PR, especially around cleaning the logic in the `BillingInfo` model. There's also a change to move the account billing endpoint to the billing namespace. The changes as I think to summarize them are as follows:

1. Limit the number of invoices being returned from Stripe by only querying for those invoice statuses we're actually interested in displaying, namely, by excluding `void` invoices. This is done starting on [StripePaymentService:1863](https://github.com/bitwarden/server/pull/4143/files#diff-3fb6472807e7c0afb6273cd35af17ae1c3d056198d60a30547d39e177d0ce7d5R1863) through to line `1889` by splitting the call to Stripe into three separate calls. An attempt was made to leave this as one call, but Stripe limits searches to a single status, forbidding arrays, so it is what it is.
2. Additionally, I added an optional parameter to limit the results returned from Stripe in [StripePaymentService:1854](https://github.com/bitwarden/server/pull/4143/files#diff-3fb6472807e7c0afb6273cd35af17ae1c3d056198d60a30547d39e177d0ce7d5R1854).
3. Similarly, I limited the number of transactions returned from the database by adding another optional parameter to [ITransactionRepository:GetManyByUserIdAsync](https://github.com/bitwarden/server/pull/4143/files#diff-05c906f3b76488963c76c4e1dac89144447400b85af832f0708d03b7cdb8c9c9R8), as well as the other two entity IDs in that file. See the migrator scripts in the changes to see that parameter added to the Transaction read stored procs, [`[dbo].[Transaction_ReadByOrganizationId]`](https://github.com/bitwarden/server/pull/4143/files#diff-97f6ed11cc35504b9e2e87a7505de5677ff9da1ad04c98f3892d04af23e2df42R1) as an example.
4. A final area of improvement was to split the `BillingInfo` object, which previously contained [`Balance`](https://github.com/bitwarden/server/pull/4143/files#diff-6b1ce01a5c123d4550039c4cd59e5a8c5eab369eef96e07657c52cc6b853b78eL9), [`PaymentSource`](https://github.com/bitwarden/server/pull/4143/files#diff-6b1ce01a5c123d4550039c4cd59e5a8c5eab369eef96e07657c52cc6b853b78eL10), [`Invoices`](https://github.com/bitwarden/server/pull/4143/files#diff-6b1ce01a5c123d4550039c4cd59e5a8c5eab369eef96e07657c52cc6b853b78eL11), and [`Transactions`](https://github.com/bitwarden/server/pull/4143/files#diff-6b1ce01a5c123d4550039c4cd59e5a8c5eab369eef96e07657c52cc6b853b78eL12). I split `Invoices` and `Transactions` into their own model, [`BillingHistoryInfo`](https://github.com/bitwarden/server/pull/4143/files#diff-955ea04bd923e9144588f99e1b8615fa77ba40171aaf99851ba215aa8e10849fR9). Later, these lists will be server-side paginated after that story goes through design.

Additionally, I did some cleanup/refactoring. I moved some unowned files into the Billing namespaces. I also split multiple classes that were in single files into their own files. I also cleaned up some of the string manipulation in [`BillingInfo`](https://github.com/bitwarden/server/pull/4143/files#diff-26a8100cf7ae3c9e503855d83f915f54e64c30fef7c658f072f6b2c9e290bc22R24), which previously relied on [`string.Concat`](https://github.com/bitwarden/server/pull/4143/files#diff-6b1ce01a5c123d4550039c4cd59e5a8c5eab369eef96e07657c52cc6b853b78eL25). Same thing with the [PayPal bit](https://github.com/bitwarden/server/pull/4143/files#diff-6b1ce01a5c123d4550039c4cd59e5a8c5eab369eef96e07657c52cc6b853b78eL77) which now simply calls [`string.PadLeft`](https://github.com/bitwarden/server/pull/4143/files#diff-26a8100cf7ae3c9e503855d83f915f54e64c30fef7c658f072f6b2c9e290bc22R68) to pad the month with zeros.

Client-side PR: https://github.com/bitwarden/clients/pull/9443

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: util/Migrator/DbScripts/2024-05-30_00_OrganizationTransactionsReadImprovements.sql
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-05-30_00_OrganizationTransactionsReadImprovements.sql:1]: You can use the CREATE OR ALTER syntax for this:

CREATE OR ALTER  PROCEDURE [dbo].[Transaction_ReadByOrganizationId]
    @OrganizationId UNIQUEIDENTIFIER,
    @Limit INT

File: util/Migrator/DbScripts/2024-05-30_02_ProviderTransactionsReadImprovements.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
+IF OBJECT_ID('[dbo].[Transaction_ReadByProviderId]') IS NOT NULL
```
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-05-30_02_ProviderTransactionsReadImprovements.sql:1]: Also here with CREATE OR ALTER syntax

File: util/Migrator/DbScripts/2024-05-30_01_UserTransactionsReadImprovements.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
+IF OBJECT_ID('[dbo].[Transaction_ReadByUserId]') IS NOT NULL
```
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-05-30_01_UserTransactionsReadImprovements.sql:1]: Also here with CREATE OR ALTER syntax

**Review by rkac-bw — DISMISSED** [review]
LGTM

**Review by cyprain-okeke — DISMISSED** [review]
Looks Good

**Review by amorask-bitwarden — DISMISSED** [review]

**Review by vincentsalucci — DISMISSED** [review]

**Comment by cturnbull-bitwarden** [comment]
@rkac-bw @amorask-bitwarden @vincentsalucci Could I get a re-review on this one when you get a chance? Had to resolve a conflict

**Review by amorask-bitwarden — APPROVED** [review]

**Review by vincentsalucci — DISMISSED** [review]

**Review by rkac-bw — APPROVED** [review]
lgtm

**Review by amorask-bitwarden — APPROVED** [review]

**Review by vincentsalucci — APPROVED** [review]

**Pull request merged**